### 1. HAPMAP DATA PROCESSING

# Download & extract HM3 data
wget ftp://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/hapmap3/plink_format/draft_2/hapmap3_r2_b36_fwd.consensus.qc.poly.map.bz2
wget ftp://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/hapmap3/plink_format/draft_2/hapmap3_r2_b36_fwd.consensus.qc.poly.ped.bz2
bunzip2 hapmap3_r2_b36_fwd.consensus.qc.poly.map.bz2
bunzip2 hapmap3_r2_b36_fwd.consensus.qc.poly.ped.bz2

# HM3 is in (old) Hg18 build.
# We want to match it with (new) 1000G data in Hg38 build.
# Therefore, we lift HM3 from Hg18 to Hg38.
# Commands based on https://shicheng-guo.github.io/bioinformatics/2017/08/01/hapmap3
# But corrected, beware of crucial differences with original commands

# Download database and script for lift
wget http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver
chmod +x liftOver
wget http://hgdownload.soe.ucsc.edu/goldenPath/hg18/liftOver/hg18ToHg38.over.chain.gz
wget https://raw.githubusercontent.com/Shicheng-Guo/GscPythonUtility/master/liftOverPlink.py

# Lift; takes some time
python2 liftOverPlink.py -m hapmap3_r2_b36_fwd.consensus.qc.poly.map -p hapmap3_r2_b36_fwd.consensus.qc.poly.ped -o hapmap3_hg38 -c hg18ToHg38.over.chain.gz -e ./liftOver

# Recode to binary format
plink --file hapmap3_hg38 --make-bed --autosome --geno 0 --allow-extra-chr --out hm3_hg38

# Add population phenotypes to plink file (abusing paternal ID column)
wget ftp://ftp.ncbi.nlm.nih.gov/hapmap/genotypes/hapmap3/plink_format/draft_2/relationships_w_pops_121708.txt
mv hm3_hg38.fam hm3_hg38.fam.org
awk 'FNR==NR { a[$1]=$7; next; } { $3 = a[$1]; print $0; }' relationships_w_pops_121708.txt hm3_hg38.fam.org > hm3_hg38.fam

### 2. 1000G DATA PROCESSING

# Download 1000G chr2 only (because we like chr2)
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/supporting/GRCh38_positions/ALL.chr2_GRCh38.genotypes.20170504.vcf.gz

# There is sample overlap between 1000G and HM3:
# https://www.internationalgenome.org/faq/does-1000-genomes-project-use-hapmap-data
# Therefore, we extract three populations that were not in HM3
# Populations based on https://www.internationalgenome.org/category/phenotype/
# and ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_sample_info.xlsx
# The commands below will give warnings about missing samples

# Extract ESN - Esan in Nigeria
bcftools view -Oz --force-samples -s HG02922,HG02923,HG02924,HG02938,HG02939,HG02941,HG02942,HG02943,HG02944,HG02945,HG02946,HG02947,HG02948,HG02952,HG02953,HG02954,HG02964,HG02965,HG02966,HG02968,HG02969,HG02970,HG02971,HG02972,HG02973,HG02974,HG02975,HG02976,HG02977,HG02978,HG02979,HG02980,HG02981,HG03099,HG03100,HG03101,HG03103,HG03104,HG03105,HG03107,HG03108,HG03109,HG03110,HG03111,HG03112,HG03113,HG03114,HG03115,HG03116,HG03117,HG03118,HG03119,HG03120,HG03121,HG03122,HG03123,HG03124,HG03125,HG03126,HG03127,HG03128,HG03129,HG03130,HG03131,HG03132,HG03133,HG03134,HG03135,HG03136,HG03137,HG03139,HG03140,HG03157,HG03158,HG03159,HG03160,HG03161,HG03162,HG03163,HG03164,HG03166,HG03167,HG03168,HG03169,HG03170,HG03171,HG03172,HG03173,HG03175,HG03176,HG03189,HG03190,HG03191,HG03193,HG03194,HG03195,HG03196,HG03197,HG03198,HG03199,HG03200,HG03202,HG03203,HG03265,HG03266,HG03267,HG03268,HG03269,HG03270,HG03271,HG03272,HG03279,HG03280,HG03291,HG03293,HG03294,HG03295,HG03296,HG03297,HG03298,HG03299,HG03300,HG03301,HG03302,HG03303,HG03304,HG03305,HG03306,HG03307,HG03308,HG03309,HG03310,HG03311,HG03312,HG03313,HG03314,HG03339,HG03341,HG03342,HG03343,HG03344,HG03350,HG03351,HG03352,HG03354,HG03361,HG03362,HG03363,HG03365,HG03366,HG03367,HG03368,HG03369,HG03370,HG03371,HG03372,HG03373,HG03374,HG03493,HG03499,HG03508,HG03510,HG03511,HG03513,HG03514,HG03515,HG03516,HG03517,HG03518,HG03519,HG03520,HG03521,HG03522 ALL.chr2_GRCh38.genotypes.20170504.vcf.gz > ESN.chr2_GRCh38.genotypes.20170504.vcf.gz &

# Extract CHS - Southern Han Chinese
bcftools view -Oz --force-samples -s HG00403,HG00404,HG00405,HG00406,HG00407,HG00408,HG00409,HG00410,HG00411,HG00418,HG00419,HG00420,HG00421,HG00422,HG00423,HG00427,HG00428,HG00429,HG00436,HG00437,HG00438,HG00442,HG00443,HG00444,HG00445,HG00446,HG00447,HG00448,HG00449,HG00450,HG00451,HG00452,HG00453,HG00457,HG00458,HG00459,HG00463,HG00464,HG00465,HG00472,HG00473,HG00474,HG00475,HG00476,HG00477,HG00478,HG00479,HG00480,HG00500,HG00501,HG00502,HG00512,HG00513,HG00514,HG00524,HG00525,HG00526,HG00530,HG00531,HG00532,HG00533,HG00534,HG00535,HG00536,HG00537,HG00538,HG00542,HG00543,HG00544,HG00556,HG00557,HG00558,HG00559,HG00560,HG00561,HG00565,HG00566,HG00567,HG00577,HG00578,HG00579,HG00580,HG00581,HG00582,HG00583,HG00584,HG00585,HG00589,HG00590,HG00591,HG00592,HG00593,HG00594,HG00595,HG00596,HG00597,HG00598,HG00599,HG00600,HG00607,HG00608,HG00609,HG00610,HG00611,HG00612,HG00613,HG00614,HG00615,HG00619,HG00620,HG00621,HG00622,HG00623,HG00624,HG00625,HG00626,HG00627,HG00628,HG00629,HG00630,HG00631,HG00632,HG00633,HG00634,HG00635,HG00636,HG00650,HG00651,HG00652,HG00653,HG00654,HG00655,HG00656,HG00657,HG00658,HG00662,HG00663,HG00664,HG00671,HG00672,HG00673,HG00674,HG00675,HG00676,HG00683,HG00684,HG00685,HG00689,HG00690,HG00691,HG00692,HG00693,HG00694,HG00698,HG00699,HG00700,HG00701,HG00702,HG00703,HG00704,HG00705,HG00706,HG00707,HG00708,HG00709,HG00716,HG00717,HG00718,HG00728,HG00729,HG00730 ALL.chr2_GRCh38.genotypes.20170504.vcf.gz > CHS.chr2_GRCh38.genotypes.20170504.vcf.gz &

# Extract FIN - Finnish in Finland
bcftools view -Oz --force-samples -s HG00171,HG00173,HG00174,HG00176,HG00177,HG00178,HG00179,HG00180,HG00181,HG00182,HG00183,HG00185,HG00186,HG00187,HG00188,HG00189,HG00190,HG00266,HG00267,HG00268,HG00269,HG00270,HG00271,HG00272,HG00273,HG00274,HG00275,HG00276,HG00277,HG00278,HG00280,HG00281,HG00282,HG00284,HG00285,HG00288,HG00290,HG00302,HG00303,HG00304,HG00306,HG00308,HG00309,HG00310,HG00311,HG00312,HG00313,HG00315,HG00318,HG00319,HG00320,HG00321,HG00323,HG00324,HG00325,HG00326,HG00327,HG00328,HG00329,HG00330,HG00331,HG00332,HG00334,HG00335,HG00336,HG00337,HG00338,HG00339,HG00341,HG00342,HG00343,HG00344,HG00345,HG00346,HG00349,HG00350,HG00351,HG00353,HG00355,HG00356,HG00357,HG00358,HG00359,HG00360,HG00361,HG00362,HG00364,HG00365,HG00366,HG00367,HG00368,HG00369,HG00371,HG00372,HG00373,HG00375,HG00376,HG00377,HG00378,HG00379,HG00380,HG00381,HG00382,HG00383,HG00384  ALL.chr2_GRCh38.genotypes.20170504.vcf.gz > FIN.chr2_GRCh38.genotypes.20170504.vcf.gz

# We run the above commands in parallel - wait for them to finish
wait

### 3. Run QGS on HM3 three times using all three reference panels

# Download gene annotation
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_35/gencode.v35.basic.annotation.gtf.gz

# Run QGS
for pop in "ESN" "CHS" "FIN"
do
  qgs --chr 2 --sample hm3_hg38.bed --reference ${pop}.chr2_GRCh38.genotypes.20170504.vcf.gz --genes gencode.v35.basic.annotation.gtf.gz --gtf-filter type=gene gene_type=protein_coding --out hm3_${pop}_ref.csv
done

### 4. Create the plots in R

mkdir -p plots/
Rscript qgs_distance.R
